safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,214 @@
+	MaxAge int
+}
+
+// Default creates a CORS Interceptor with default settings.
+// Those defaults are:
+//  - No Exposed Headers
+//  - No Allowed Headers
+//  - AllowCredentials: false
+//  - MaxAge: 5 seconds
+func Default(allowedOrigins ...string) *Interceptor {
+	ao := map[string]bool{}
+	for _, o := range allowedOrigins {
+		ao[o] = true
+	}
+	return &Interceptor{
+		AllowedOrigins: ao,
+		allowedHeaders: map[string]bool{},
+	}
+}
+
+// SetAllowedHeaders sets which headers should be allowed in the Access-Control-Allow-Headers
+// header. The headers are first canonicalized using textproto.CanonicalMIMEHeaderKey.
+func (it *Interceptor) SetAllowedHeaders(headers ...string) {
+	it.allowedHeaders = map[string]bool{}
+	for _, h := range headers {
+		if h == "*" {
+			continue
+		}
+		it.allowedHeaders[textproto.CanonicalMIMEHeaderKey(h)] = true
+	}
+}
+
+// Before handles the incoming request and sets headers or responds with an error
+// accordingly.
+func (it *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	origin := r.Header.Get("Origin")
+	if origin != "" && !it.AllowedOrigins[origin] {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+	h := w.Header()
+	allowOrigin, err := h.Claim("Access-Control-Allow-Origin")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	vary, err := h.Claim("Vary")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	allowCredentials, err := h.Claim("Access-Control-Allow-Credentials")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	var status safehttp.StatusCode
+	if r.Method() == safehttp.MethodOptions {
+		status = it.preflight(w, r)
+	} else {
+		status = it.request(w, r)
+	}
+
+	if status != 0 && status != safehttp.StatusNoContent {
+		if 400 <= status && status < 500 {
+			return w.ClientError(status)
+		} else {
+			return w.ServerError(status)
+		}
+	}
+
+	if origin != "" {
+		allowOrigin([]string{origin})
+		vary([]string{"Origin"})
+	}
+	if r.Header.Get("Cookie") != "" && it.AllowCredentials {
+		// TODO: handle other credentials than cookies:
+		// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+		allowCredentials([]string{"true"})
+	}
+
+	if status == safehttp.StatusNoContent {
+		return w.NoContent()
+	}
+	return safehttp.Result{}
+}
+
+// preflight handles requests that have the method OPTIONS.
+func (it *Interceptor) preflight(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.StatusCode {
+	rh := r.Header
+	method := rh.Get("Access-Control-Request-Method")
+	if method == "" {
+		return safehttp.StatusForbidden
+	}
+	wh := w.Header()
+	allowMethods, err := wh.Claim("Access-Control-Allow-Methods")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	headers := rh.Get("Access-Control-Request-Headers")
+	if headers != "" {
```

Comment:
This `if` is not needed.

Reply:
Yes, it is. Running `strings.Split("", ", ")` gives a slice with a single empty string as an element (`[]string{""}`) which will make the loop run and will cause a 403 Forbidden response since `""` isn't an allowed header. So either we have this check or we add `""` as an allowed header.